fix: Restore detection of letsencrypt certificate file changes

polarathene opened the pull request:
# Description

The `DYNAMIC_FILES` var was quote wrapped, treating all filepaths to create checksums for as a single string that would be ignored instead of processed individually.

Removed the quotes, and changed the for loop to an array which accomplishes the same goal.

This was discovered while investigating [#2321](https://github.com/docker-mailserver/docker-mailserver/issues/2321#issuecomment-994491231), but it's unclear if this will fix that issue.

---

This bug was introduced in https://github.com/docker-mailserver/docker-mailserver/pull/2279 with the `v10.3.0` release.

~~We may want to branch that release tag and cherry pick this commit from master to publish a bugfix (_as we've already merged `v11` changes to master, the bugfix branch is the only way to backport it I think?_).~~ (**RESOLVED:** Decision was to release in `v10.4.0` with the Debian upgrade)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

It'd probably be a good idea to add a test that verifies the cert path is in the checksum file, I'm just a bit pressed on time and resources currently but wanted to get this fix raised while I could spare the time.

If there's no rush to merge I can sort out tests within 7 days depending on how events pan out on my end :sweat_smile:

georglauterbach commented:
I feel like it would be a good idea for you to run

``` BASH
cat >.git/hooks/pre-commit << "EOM"
#! /bin/bash

set -euEo pipefail
make lint
EOM
```

to setup a pre-commit hook that does the linting :D

georglauterbach commented:
I've never done a cherry pick but I'd like to learn it :D From what you said, this looks like a valuable solution for a bugfix backport.

wernerfred commented:
> We may want to branch that release tag and cherry pick this commit from master to publish a bugfix (_as we've already merged `v11` changes to master, the bugfix branch is the only way to backport it I think?_)

IIRC the release CI doesn't trigger on other branches than master so when cherry picked you might need to build and push the image manually to the registries.

This problem follows up the discussion in the maintainer team on the git strategy. Our current setup is not set up to support backports. Implementing such support should be weigh up against the need of backports (didn't happen often in the past, at least since i participate). Another (but ugly) way would be to revert the `v11` commit merge and publish this fix and then add the `v11` changes again

casperklein commented:
> other (but ugly) way would be to revert the v11 commit merge and publish this fix and then add the v11 changes again

Ugly yes. But easy to do and resulting in very small changes to revert.
I've some PR on my to-do, but wanted to wait, until it's clear, that the next release will definitely be v11 and not another 10.x. 

I am a bit confused wether we treat Debian11 as breaking change and therefore a requirement for our v11 release or not. My last impression was, it's not related to our v11 release.

wernerfred commented:
> My last impression was, it's not related to our v11 release.

If thats the case we need no revert and can merge straight away... 🚀

polarathene commented:
> IIRC the release CI doesn't trigger on other branches than master so when cherry picked you might need to build and push the image manually to the registries.

My recall is that tags are not associated to branches, so as long as we have CI that works on the version tags like the docs deployment does, that should work fine via CI?

> This problem follows up the discussion in the maintainer team on the git strategy. Our current setup is not set up to support backports.

I remember that discussion and what I proposed here is exactly what I thought I had suggested was the agreed approach for handling this sort of fix?

> Another (but ugly) way would be to revert the v11 commit merge and publish this fix and then add the v11 changes again

We have done that in the past, the only issue without the bugfix is that anyone running `10.3.0` without `acme.json` for cert management will not get their certs automatically detected upon renewal.

In my own local testing, this did not seem to prevent Postfix from using the renewed cert, even though Postfix was not restarted. I did not test Dovecot however, and I'm not sure why the linked issue was experiencing a problem, so it may depend on how the cert files are provided too.

> I am a bit confused wether we treat Debian11 as breaking change and therefore a requirement for our v11 release or not. My last impression was, it's not related to our v11 release.

Well it did cause breaking changes originally that needed to be fixed. It also upgrades Postfix/Dovecot and has that potential Fail2Ban CVE. Seems enough to warrant being part of a major release, although I wanted to delay that release until sometime in Jan as I was wanting to make other notable changes (_eg: restoring `flock()` file locking, and refactoring SSL/TLS logic to unify modes, rename `intermediate` TLS profile to `legacy`_).

---

If you think it's better to skip a `10.3.1` to merge the Debian 11 as `10.4.0` that's fine, but we should update release notes/changelog about `10.3.0` with a warning that the change detection doesn't react to certificate changes in `/etc/letsencrypt`?

I don't think the Debian 11 upgrade warrants being part of a patch release, only a minor or major.

polarathene commented:
> to setup a pre-commit hook that does the linting :D

This wouldn't work when making a commit via the Github UI right? My computer isn't presently able to do much work without risking an OOM event lately, so I just contribute small PRs via the web interface and do any testing/verification via a VPS instance with SSH.

Pre-commit wouldn't help with the VPS unless I get into the habit of committing and pushing in the VPS too, but I could at least get into the habit of running the lint checks before opening a PR :sweat_smile: The CI check just lets me be lazy and address it afterwards (since we squash merge commit history :stuck_out_tongue: )

> I've never done a cherry pick but I'd like to learn it :D From what you said, this looks like a valuable solution for a bugfix backport.

I don't know if I do it right whenever I've done it. I have only done so via GitKraken UI, I can right click on a commit from another branch and cherry pick it over, I'm not entirely sure if that is any different than just copy/pasting the change (_just with a fancy git command to do copy/paste via commits instead_).

---

It looks like this will trigger CI just fine via `v10.3.1` tag, doesn't need to be on master branch:

https://github.com/docker-mailserver/docker-mailserver/blob/6ad9dd30635a6e81dfe8dcecb933ee445a1cb145/.github/workflows/default_on_push.yml#L1-L15

wernerfred commented:
> 1. as a Christmas gift for sysadmins to complain about

😆 😆 😆

polarathene commented:
> 1. just merging this on top, (waiting one more week and) releasing 10.4 (as a Christmas gift for sysadmins to complain about) and calling it a day (but feeling like Santa himself)

I guess this is the easiest choice if we're not going to try the cherrypick.

I should add the changelog update for `10.3.0` in this PR and update the release page to both add a warning about the regression?

---

It looks like this change also broke some related tests, joy.. :sweat_smile: 

I'll need to investigate and resolve those first.

polarathene commented:
Tests resolved, fixed another bug (_didn't previously affect any tests as the method was only ever used with `SSL_DOMAIN` to test against documented wildcard prefix_).

Refactored the `check-for-changes.sh` logic for `acme.json` a bit by removing the unnecessary for loop and switch statement. Just uses a regex conditional now. View diff with white-space hidden to reduce noise.

- Cert files extracted for `acme.json` now have their hashes re-calculated afterwards to avoid the redundant change event that they otherwise triggered. Using `sha512sum` is run without subshell + exec + stderr redirect, I think that's fine?
- The helper method `_strip_wildcard_prefix` now returns the original input if it didn't have a wildcard prefix to remove. Without that, the usage with `CERT_DOMAIN` and no `SSL_DOMAIN` filtering caused the logic to hang, likewise if `SSL_DOMAIN` with a wildcard was not stripped. I assume something went wrong with `sed` or `sha512sum` with the `*.`.
- Change detection test increased the amount of service log checked, the amount (3000 bytes) is matching an existing value used elsewhere in that test file which is more than enough.

---

The two tests passed locally, I expect all tests to be good now :+1: 

I'll add the warning for `10.3.0` release not reacting to cert file changes and prep a `10.4.0` release.

georglauterbach commented on target/scripts/helper-functions.sh:
```diff
@@ -187,9 +187,9 @@ function _extract_certs_from_acme
+# Remove the `*.` prefix if it exists, else returns the input value
 function _strip_wildcard_prefix {
-  [[ "${1}" == "*."* ]] && echo "${1:2}"
+  [[ "${1}" == "*."* ]] && echo "${1:2}" || echo "${1}"
```
More than surprised ShellCheck did not pick this up. IIRC in Bash, this is not an if-then-else. But I will look that up.

casperklein replied:
The second `echo `is not executed, if the first `echo `fails. That's why you thought, it's not a real if-statement.

I think in that case, the shorthand is fine, as echo is unlikely to fail.

georglauterbach replied:
I guess it's fine in this case, right.

casperklein replied:
I've edited my comment above. It's the opposite of what I wrote 😖

georglauterbach reviewed (approved):
LGTM 👍 

I will have a look at the if-then-else thing today.
